trio/tests/test_exports.py

```diff
@@ -114,6 +143,169 @@ def no_underscores(symbols):
         assert False
 
 
+# this could be sped up by only invoking mypy once per module, or even once for all
+# modules, instead of once per class.
+@slow
+# see commend on test_static_tool_sees_all_symbols
+@pytest.mark.redistributors_should_skip
+# pylint/jedi often have trouble with alpha releases, where Python's internals
+# are in flux, grammar may not have settled down, etc.
+@pytest.mark.skipif(
+    sys.version_info.releaselevel == "alpha",
+    reason="skip static introspection tools on Python dev/alpha releases",
+)
+@pytest.mark.parametrize("module_name", PUBLIC_MODULE_NAMES)
+@pytest.mark.parametrize("tool", ["jedi", "mypy"])
+def test_static_tool_sees_class_members(tool, module_name, tmpdir) -> None:
+    module = PUBLIC_MODULES[PUBLIC_MODULE_NAMES.index(module_name)]
+
+    # ignore hidden, but not dunder, symbols
+    def no_hidden(symbols):
+        return {
+            symbol
+            for symbol in symbols
+            if (not symbol.startswith("_")) or symbol.startswith("__")
+        }
+
+    if tool == "mypy":
+        if sys.implementation.name != "cpython":
+            pytest.skip("mypy not installed in tests on pypy")
+        # create py.typed file
+        (Path(trio.__file__).parent / "py.typed").write_text("")
+
+    errors: dict[str, Any] = {}
+    if module_name == "trio.tests":
+        return
+    for class_name, class_ in module.__dict__.items():
+        if not isinstance(class_, type):
+            continue
+        if module_name == "trio.socket" and class_name in dir(stdlib_socket):
+            continue
+        # Deprecated classes are exported with a leading underscore
+        if class_name.startswith("_"):  # pragma: no cover
```

Comment:
Given the `pragma: no cover`, am I right to assume this doesn't happen at the moment? I presume if we ever do export a deprecated class (?), it'll have member annotations so this exception might be unnecessary.

Reply:
MultiError is exported as `_MultiError`, so it is happening at the moment. But I didn't wanna care about whether code that's on it's way out is typed correctly or not, hence skipping it.
The reason for the `pragma` is so that when `_MultiError` *does* get removed it doesn't decrease coverage in this random test.

I think that was my reasoning at least, I'll make the comment clearer

Reply:
Oh, except `MultiError` has no problems anyway (now, it got fixed with all the other exception fixes) - so I'll just remove the if/continue.

Reply:
nvm, it does have problems https://github.com/jakkdl/trio/actions/runs/4893133337/jobs/8735733831#step:5:1779
reintroduced the if/continue and clarified the comment